subsys/net/ip/tp.c

```diff
@@ -448,14 +424,36 @@ static void tp_pkt_send(struct net_pkt *pkt)
 	tp_pkt_unref(pkt, tp_basename(__FILE__), __LINE__);
 }
 
+struct net_pkt *tp_output_pkt_alloc(sa_family_t af, struct net_if *iface,
+				    const char *file, int line)
+{
+	struct tp_pkt *tp_pkt = k_malloc(sizeof(struct tp_pkt));
+
+	tp_assert(tp_pkt, "");
+
+	tp_pkt->pkt = net_pkt_alloc_with_buffer(iface,
+						sizeof(struct net_udp_hdr),
+						af, IPPROTO_UDP, K_NO_WAIT);
+	tp_assert(tp_pkt->pkt, "");
+
+	tp_pkt->file = file;
+	tp_pkt->line = line;
+
+	sys_slist_append(&tp_pkts, (sys_snode_t *) tp_pkt);
+
+	return tp_pkt->pkt;
+}
+
 void _tp_output(sa_family_t af, struct net_if *iface, void *data,
 		size_t data_len, const char *file, int line)
 {
-	struct net_pkt *pkt = tp_pkt_alloc(0, file, line);
+	struct net_pkt *pkt = tp_output_pkt_alloc(af, iface, file, line);
```

Comment:
you know the data length, go ahead: pass it to tp_output_pkt_alloc() and make it allocate for a udp hdr + the data length.